api/samples/neoncounter.c

```diff
@@ -0,0 +1,974 @@
+                     uint num_simd_store, uint num_branching, uint num_load,
+                     uint num_load_linear, uint num_load_structured,
+                     uint num_store, uint num_store_linear,
+                     uint num_store_structured, uint num_sve,
+                     uint num_sve_arith, uint num_sve_load,
+                     uint num_sve_store) {
+
+  count_all += num_instrs;
+  count_arith += num_arith;
+  count_simd += num_simd;
+  count_simd_arith += num_simd_arith;
+  count_simd_load += num_simd_load;
+  count_simd_store += num_simd_store;
+  count_branching += num_branching;
+  count_load += num_load;
+  count_load_linear += num_load_linear;
+  count_load_structured += num_load_structured;
+  count_store += num_store;
+  count_store_linear += num_store_linear;
+  count_store_structured += num_store_structured;
+  count_sve += num_sve;
+  count_sve_arith += num_sve_arith;
+  count_sve_load += num_sve_load;
+  count_sve_store += num_sve_store;
+}
+
+// in DEBUG mode all processed instructions are written in this file
+FILE *debug;
+
+// remembers if only the app or the app and all shared libraries should be
+// analyzed
+static bool app_only = false;
+// application module
+static app_pc exe_start;
+
+// saves the name of the executed app
+static const char *executable;
+
+/*
+ * analyzes an instruction list
+ */
+void analyze_instr(instr_t *instr, void **user_data) {
+  // save counts for all instr., simd instr. and branching instr.
+  uint *counts = malloc(sizeof(uint) * 17);
+  // all instrs.
+  counts[0] = 0;
+  // all arithmethic count
+  counts[1] = 0;
+
+  // neon
+  counts[2] = 0;
+  // neon arithmetic
+  counts[3] = 0;
+  // neon load
+  counts[4] = 0;
+  // neon store
+  counts[5] = 0;
+  // branching
+  counts[6] = 0;
+
+  // load
+  counts[7] = 0;
+  // load linear
+  counts[8] = 0;
+  // load structured
+  counts[9] = 0;
+
+  // store
+  counts[10] = 0;
+  // store linear
+  counts[11] = 0;
+  // store structured
+  counts[12] = 0;
+
+  // sve
+  counts[13] = 0;
+  // sve arithmetic
+  counts[14] = 0;
+  // sve load
+  counts[15] = 0;
+  // sve store
+  counts[16] = 0;
+
+  const char *op_name;
+  int op_code;
+  // remember if instruction is load/store instruction
+  bool is_load = false;
+  bool is_store = false;
+  // remember if instruction is arithmetic instruction
+  bool is_ar = false;
+  bool is_neon = false;
+  bool is_sve = false;
+
+  instr_t *ins;
+
+  for (ins = instr; ins != NULL; ins = instr_get_next(ins)) {
+    // reset
+    is_load = false;
+    is_store = false;
+    is_ar = false;
+    is_neon = false;
+
+    if (drmgr_is_emulation_end(ins)) {
+      is_sve = false;
+      continue;
+    }
+    if (is_sve) {
+      /* An SVE instruction is emulated with general AArch64
+       * instructions, but we only want to count it once. */
+      continue;
+    }
+    if (drmgr_is_emulation_start(ins)) {
+      //
+      is_sve = true;
```

Comment:
This is only the case if the emulation client is for SVE, the emulation API isn't tied to any specific architecture or extension. All that can be known here without looking at the encoding is the instruction is emulated (see [1]). Until SVE is supported natively in DynamoRIO this is tricky, I think it's worth removing SVE counting from this client and keeping it to just NEON for now.

[1] https://github.com/DynamoRIO/dynamorio/blob/master/api/samples/inscount.cpp#L168